### PR TITLE
feat(example-greeting-app): add an example greeting app on top of greeter extension

### PR DESCRIPTION
**Pull request by raymondfeng**
Introducing a new `greeting-app` example on top of  `greeter-extension` to prepare for a series of tutorials #2939:
 
1. Add REST API for greeting service
2. Show case caching with interceptor and life cycle observer
3. Use `@config.*` and `ctx.configure`

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: examples/greeter-extension/index.js [hunk 1]
```diff
@@ -5,7 +5,18 @@
...
+    const cli = await app.main();
+    if (!cli) {
+      console.log(
+        `The service is running at ${app.restServer.url}/greet/world`,
+      );
+    }
+  }
 }
+
+main();
```
Comment (bajtos) [inline examples/greeter-extension/index.js]: I find it suspicious that `main` is always executed. 

I believe we are using the following approach in other places:

```ts
if (require.main === module) {
  main().catch(err => {
    console.error(err);
    process.exit(1);
  });
}

async function main() {
  const app = new module.exports.GreetingApplication({
    // ...
  }
}
```

**Comment by raymondfeng** [comment]
> I am confused. Initially, example-greeter-extension was created to show how to create an extension (a LB4 Component). Now it looks more like a regular application.

I'm trying to reuse this scenario for https://github.com/strongloop/loopback-next/pull/2939. 

> How do you envision 3rd party applications to use example-greeter-extension as a Component?

The module can still be imported by another app, such as:
```ts
import {GreetingComponent} from '@loopback/example-greeter-extension';

app.component(GreetingComponent);
```

Maybe it makes sense to break the features into two packages:
1. example-greeter-extension - stay as a component
2. example-greeter-app - a new app that uses `example-greeter-extension` and adds new features such as REST APIs and caching

**Comment by raymondfeng** [comment]
FYI: I went ahead to refactor the code as follows:

1. example-greeter-extension - stay as a component
2. example-greeter-app - a new app that uses example-greeter-extension and adds new features such as REST APIs and caching

File: examples/greeting-app/src/caching-service.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+export class CachingService {
+  private timer: NodeJS.Timer;
+  private store: Map<string, Message> = new Map();
+
+  constructor(
+    @config.view()
+    private optionsView: ContextView<CachingServiceOptions>,
+  ) {
+    // Use a view so that we can listen on the `refresh` event
+    optionsView.on('refresh', async () => {
```
Comment (b-admike) [inline examples/greeting-app/src/caching-service.ts]: When does `refresh` event happen? Can we capture that in a comment?

File: examples/greeting-app/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+## Contributions
+
+- [Guidelines](https://github.com/strongloop/loopback-next/blob/master/docs/CONTRIBUTING.md)
+- [Join the team](https://github.com/strongloop/loopback-next/issues/110)
+
+## Try out
+
+1. As a command line
+
+   Run `npm start -- -c` from the root folder to run the sample application. You
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply examples/greeting-app/README.md]: You are looking at an outdated version. The instructions should be good now.

**Comment by raymondfeng** [comment]
> I would really like to see a bit more beefy README file for the new greeter app. Can we explain a bit more on the caching interceptor/observer/service? Also, we should let our users know that they need to set Accept-Language header when trying it out and modify the path /greeter/<Name> to play with it. It wouldn't hurt to mention to turn on debug strings with DEBUG=greeter-extension* e.g. to showcase what the caching observer/interceptor are doing in the background. It'd be nice to include explorer with the app if it can allow users to put in the headers for the endpoint (I know we're trying to keep this lightweight).

I'm trying to cover them in more details at https://github.com/strongloop/loopback-next/pull/2939.

**Comment by raymondfeng** [comment]
@b-admike Thank you for the feedback. PTAL.

File: examples/greeting-app/src/caching-service.ts [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    debug('Caching options: %j', options);
+    return (options && options.ttl) || 5000;
+  }
+
+  /**
+   * Remove expired items from the cache
+   */
+  async sweep() {
+    debug('Sweeping cache...');
+    for (const key of this.store.keys()) {
```
Comment (emonddr) [inline examples/greeting-app/src/caching-service.ts]: I guess the Map object in javascript is thread safe? (access to keys is locked so reads/writes cannot occur at same time as delete)

Comment (emonddr) [reply examples/greeting-app/src/caching-service.ts]: I retract my comment...javascript runs on one thread.

File: examples/greeting-app/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/example-greeting-app
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+export * from './dist';
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply examples/greeting-app/index.d.ts]: This is always the case for all packages.

File: examples/greeting-app/index.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  const app = new module.exports.GreetingApplication({
+    rest: {host: '127.0.0.1', port: 3000},
+  });
+  await app.main();
+  console.log(`The service is running at ${app.restServer.url}/greet/world`);
+}
+
+if (require.main === module) {
+  runApp();
+}
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply examples/greeting-app/index.js]: I made it more consistent.

**Pull request merged**